Js2

Queenen opened the pull request:
I would like to know if all the CA requirements are fulfilled.
Also, how to fix the following console errors:

Some cookies are misusing the recommended “SameSite“ attribute.
A resource is blocked by OpaqueResponseBlocking, please check browser console for details.

Queenen committed:
Transferring files

Index: login/register authentication form,
Feed: JS feed ( shows all posts with body content), JS searchbar for finding specific content

Queenen committed:
Filter function works

The user may click the "images" or "posts" sort options to find posts matching these criterias. Sorting options that are not available is now disabled in HTML. (polls, videos etc.)

Queenen committed:
Merge pull request #1 from Queenen/filter-posts

Filter function works

Queenen committed:
Post creation

The user may now create a post by clicking the top nav pencil icon, fill out the modal info and click "post".

Queenen committed:
Merge pull request #2 from Queenen/post-creation

Post creation

Queenen committed:
Edit/Delete post

Whenever a post is associated with the correct unique user email ID, the user may edit the post. When in edit mode the user may change the content and save changes, or delete the post entirely.

Queenen committed:
Merge pull request #3 from Queenen/edit-post

Edit/Delete post

Queenen committed:
Added Media URL

The user may now upload an image to their post using an url

Queenen committed:
Merge pull request #4 from Queenen/edit-post

Added Media URL

Queenen committed:
JS clean up

Cleaning up files, separating commonly used functions into their own JS files.

Queenen committed:
Merge pull request #6 from Queenen/clean-up

JS clean up

Queenen committed:
currentUsername replaced

Replacing currentUsername authorization with currentID email (for cases with already registered users).

Queenen committed:
Merge pull request #7 from Queenen/clean-up

currentUsername replaced

Queenen committed:
update files

Queenen committed:
Profile page

The user may now change his/her banner, avatar and profile description

Queenen committed:
Merge pull request #8 from Queenen/profile

Profile page

Queenen committed:
Profile page

Did a small change that influenced the fetching of username. I also added a loader that appears when the page is loading.

Queenen committed:
Merge pull request #9 from Queenen/profile

Profile page

Fermain reviewed (changes requested):
This is a good quality project that has one consistent problem in terms of JS and UX. This is the most common issue on this CA, Error handling.

The issue is that you are catching errors in the wrong place, lets make a simple example:

`getExample()` is the API function.
`renderExample(exampleData)` is the function that shows HTML.
`examplesPage()` is the function controlling everything on this page.

```js
import { getExample } from "somewhere";
import { renderExample } from "somewhere";

export async function examplesPage() {
  try {
    const example = await getExample();
    renderExample(example);
  } catch (error) {
    return alert("Could not access example data');
  }
}
```

In this case we need `getExample` to throw an error if something goes wrong. If this function has a try/catch, it can never throw an error, because this will always be caught.

Queenen committed:
Profile/Feed

Improved the search functionality,
and made it possible for the user to change his/her banner, avatar and profile description. The profile js also fetches and shows the users posts.

Queenen committed:
Merge pull request #10 from Queenen/profile

Profile/Feed

Queenen committed:
Profile

Did some debugging that affected the HTML of new profiles.

Queenen committed:
Merge pull request #12 from Queenen/profile

Profile

Queenen committed:
Error Handling

The error handling should be more smooth, throwing errors more often if something went wrong.

Queenen committed:
Merge branch 'js2' into error_handling

Queenen committed:
Merge pull request #13 from Queenen/error_handling

Error Handling

Queenen committed:
Profile.js

Fixed some issues with js.

Queenen committed:
Merge pull request #14 from Queenen/error_handling

Profile.js

Queenen committed:
README.md

Added a description of the project.

Queenen committed:
Merge pull request #15 from Queenen/clean-up

README.md

Queenen committed:
Update README.md

Queenen committed:
W3 and meta

some w3 changes, and added meta description for all pages.

sulenchy commented on js/create.js:
```diff
@@ -0,0 +1,71 @@
+    );
+
+    if (!result) {
+      throw new Error("Failed to create post");
```
Make sense. But this error will only show up in the browser's console. It would be better to display a notification on the UI.

Queenen replied:
The HTTP_requst_base.js provides intel to the user about what went wrong with an alert message. 
Example: If the user tries to create a post when the JWT is not present, the user is alerted with the following message: 
"Failed to create post: Request failed with status 401: No authorization header was found".

But now I've also added an inline HTML error message as prompted.

Queenen replied:
By default the user should get alerted once an error occur. But I now have added inline error message for the user as well.

sulenchy commented on js/edit_delete.js:
```diff
@@ -0,0 +1,116 @@
+      localStorage.getItem("token")
+    );
+    if (!result) {
+      throw new Error("Error editing post");
```
Same thing hee. Show a notification(error)

Queenen replied:
Should now be showing inline error messages for the user to see.

sulenchy commented on js/edit_delete.js:
```diff
@@ -0,0 +1,116 @@
+      localStorage.getItem("token")
+    );
+    if (!result) {
+      throw new Error("Error deleting post");
```
Same thing here

Queenen replied:
Should now be showing inline error messages for the user to see.

sulenchy reviewed (changes requested):
@Queenen Well done. This is a good job. However, two required features are missing, and for you to pass this CA, you need to get them in. The features are:

- User can filter the post content feed (most important).
- User can view a post content item by ID

You have an ultimatum of 24 hours to get this done. All the best

Queenen committed:
Create post error handling

Queenen committed:
Merge pull request #16 from Queenen/error_handling

Create post error handling

Queenen committed:
edit/delete error handling

added inline html message for when a bad request happens. did some slight changes to alerts.

Queenen committed:
Merge pull request #17 from Queenen/error_handling

edit/delete error handling

Queenen committed:
view post by ID

additional post page where specific posts are shown by ID.

Queenen committed:
Merge pull request #18 from Queenen/view-post-by-ID

view post by ID

Queenen committed:
improved error handling

Queenen committed:
Merge branch 'js2' into error_handling

Queenen committed:
Merge pull request #19 from Queenen/error_handling

improved error handling

Queenen committed:
more error handling
